Library/Homebrew/tap.rb

```diff
@@ -931,27 +943,42 @@ def ==(other)
     other = Tap.fetch(other) if other.is_a?(String)
     other.is_a?(self.class) && name == other.name
   end
+  alias eql? ==
 
-  def self.each(&block)
-    return to_enum unless block
+  sig { returns(Integer) }
+  def hash
+    [self.class, name].hash
+  end
 
-    installed_taps = if TAP_DIRECTORY.directory?
-      TAP_DIRECTORY.subdirs
-                   .flat_map(&:subdirs)
-                   .map(&method(:from_path))
+  # All locally installed taps.
+  sig { returns(T::Array[Tap]) }
+  def self.installed
```

Comment:
Let's keep these a private implementation detail. I don't really want to have three different ways to do the same thing, e.g. `Tap.select(&:installed?)` vs. `Tap.all.select(&:installed?)` vs.  `Tap.installed`, and `Tap.all` vs. `Tap.to_a`.

Reply:
I think it's preferable to use `Tap.installed` vs. `Tap.select(&:installed?)` which is now used widely throughout the codebase if we're already caching that value directly. `Tap.all` is currently only used in tests so I think that we can make more of an argument that it can be made private. Keep in mind that `Tap.all` and `Tap.to_a` are not always equivalent.

Reply:
To me, this is an implementation detail to make `each` faster. In order to have the same performance benefits without exposing a new method, we can do:

```ruby
  def self.select(&block)
    return installed if block == proc(&:installed?)

    super
  end
```

Reply:
I think that now that `Tap.each` returns different results depending on whether the API is being used the code becomes a bit more confusing. Why do we need to check that taps are installed or not? That is also an implementation detail. `Tap.installed` is simpler to understand when reading the codebase. It was not me intent to add a new method here; that just flowed logically from the two states that `Tap.each` can represent within the lifetime of the program. But now that it's here I think we should use it.

Is this disagreement a blocker for this PR?

Reply:
> To me, this is a nice helper function that you removed due to your personal preference

@MikeMcQuaid, exactly when did I remove `Tap::installed`?

Reply:
@reitermarkus You made https://github.com/Homebrew/brew/pull/16710 which required changes like https://github.com/Homebrew/homebrew-bundle/pull/1317.

Reply:
I see what you mean, `Tap::each` was basically `Tap::installed` before. Still, that wasn't personal preference but necessary to improve `Tap.clear_cache` behaviour and fix `Tap::reverse_tap_migrations_renames`.